src/parse.ts

```diff
@@ -0,0 +1,38 @@
+export type ParamsObject = Record<string, string | string[]>
+
+export interface ParsedURL {
+  protocol: string
+  hostname: string
+  port: string
+  username: string
+  password: string
+  pathname: string
+  hash: string
+  params: ParamsObject
+}
+
+export function parseURLNative (input: string = '', _hasProtocol: boolean): ParsedURL {
+  const url: URL & { params: ParamsObject } = new URL(input, _hasProtocol ? undefined : 'default:/') as any
+  url.params = entriesToObject(url.searchParams.entries())
+  return url
+}
+
+export function hasProtocol (inputStr: string): boolean {
+  return /^\w+:\//.test(inputStr)
+}
+
+export function entriesToObject<T = string> (entries: IterableIterator<[string, T]>): Record<string, T | T[]> {
```

Comment:
Can the generic type just be string type?

Reply:
Makes sense updated